python/llm/src/bigdl/llm/transformers/models/baichuan2.py

```diff
@@ -301,3 +301,94 @@ def baichuan_attention_forward_13b(
         attn_weights = None
 
     return attn_output, attn_weights, past_key_value
+
+
+def _get_interleave(n):
+    def _get_interleave_power_of_2(n):
+        start = 2 ** (-(2 ** -(math.log2(n) - 3)))
+        ratio = start
+        return [start * ratio**i for i in range(n)]
+
+    if math.log2(n).is_integer():
+        return _get_interleave_power_of_2(n)
+    else:
+        closest_power_of_2 = 2 ** math.floor(math.log2(n))
+        return (
+            _get_interleave_power_of_2(closest_power_of_2)
+            + _get_interleave(2 * closest_power_of_2)[0::2][: n - closest_power_of_2]
+        )
+
+
+def _fill_with_neg_inf(t):
+    """FP16-compatible function that fills a tensor with -inf."""
+    return t.float().fill_(float("-inf")).type_as(t)
+
+
+def _buffered_future_mask(tensor, maxpos, alibi, attn_heads):
+    _future_mask = torch.triu(_fill_with_neg_inf(torch.zeros([maxpos, maxpos])), 1)
+    _future_mask = _future_mask.unsqueeze(0) + alibi
+    new_future_mask = _future_mask.to(tensor)
+    return new_future_mask[: tensor.shape[0] * attn_heads, :maxpos, :maxpos]
+
+
+def baichuan_13b_gen_alibi_mask(tensor, n_head, max_pos):
+    # May use fp16 for alibi mask to further reduce memory
+    slopes = torch.Tensor(_get_interleave(n_head))#.half()
```

Comment:
For now first disable fp16 mask since it will impact a bit of the performance.
The other changes won't hurt the performance.
@jason-dai Is it OK to merge this PR first after I test on MTL?
Will investigate if memory could be further reduced in a future PR.

Reply:
Does 1K input work?

Reply:
Previously worked, but it doesn't work starting from yesterday. After emptying mask cache, it can work now, see more details here: https://github.com/analytics-zoo/nano/issues/815#issuecomment-1866193709